bcda/auth/middleware.go

```diff
@@ -6,43 +6,61 @@ import (
 	"fmt"
 	"net/http"
 	"regexp"
-	"strings"
 
 	"github.com/CMSgov/bcda-app/bcda/responseutils"
 	jwt "github.com/dgrijalva/jwt-go"
 	"github.com/pborman/uuid"
 	log "github.com/sirupsen/logrus"
 )
 
-func RequireTokenAuth(next http.Handler) http.Handler {
+func ParseToken(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		authHeader := r.Header.Get("Authorization")
-
 		if authHeader == "" {
-			log.Error("no token in header")
-			respond(w, http.StatusUnauthorized)
+			next.ServeHTTP(w, r)
 			return
 		}
 
-		tokenString := strings.Split(authHeader, " ")[1]
-		token, err := GetProvider().DecodeJWT(tokenString)
+		authRegexp := regexp.MustCompile(`^Bearer (\S+)$`)
+		authSubmatches := authRegexp.FindStringSubmatch(authHeader)
+		if len(authSubmatches) < 2 {
+			log.Warn("Invalid Authorization header value")
+			next.ServeHTTP(w, r)
+			return
+		}
 
+		tokenString := authSubmatches[1]
+		token, err := GetProvider().DecodeJWT(tokenString)
 		if err != nil {
-			log.Error(err)
-			respond(w, http.StatusUnauthorized)
+			log.Error(err.Error())
+			next.ServeHTTP(w, r)
 			return
 		}
 
-		err = GetProvider().ValidateJWT(tokenString)
+		ctx := context.WithValue(r.Context(), "token", &token)
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
 
-		if err != nil {
-			log.Error(err)
+func RequireTokenAuth(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		token := r.Context().Value("token")
+		if token == nil {
+			log.Error("No token found")
 			respond(w, http.StatusUnauthorized)
 			return
 		}
 
-		ctx := context.WithValue(r.Context(), "token", &token)
-		next.ServeHTTP(w, r.WithContext(ctx))
+		if token, ok := token.(*jwt.Token); ok {
+			err := GetProvider().ValidateJWT(token.Raw)
```

Comment:
Yeah, that was bugging me too. A simple alternative would be to keep everything in `RequireTokenAuth`, but that would mean putting that and the logging middleware on every route. Maybe a version of the validate function that takes the already-decoded `Token` instead of the string value?

Reply:
>Maybe a version of the validate function that takes the already-decoded `Token` instead of the string value?
thats what i was thinking.  and the current validate would just decode and then call this new one.  seems out of scope for here but could be an optimization later

Reply:
Ticket to review the token middlewares for improvements: https://jira.cms.gov/browse/BCDA-927

Reply:
The future optimization should be that ParseToken() decodes the token into an auth.Token. That way, we can normalize handling of either alpha or okta tokens, and consistently pass auth.Token in the Provider interface.